File: day015_pure_recursion/related_print_n_to_1/solution.java
import java.util.Scanner;

class Solution {
    // Optimized (1 to N): Post-recursion print
    public static void print_1_to_N_optimal(int n) {
        if (n == 0) {
            return;
        }
        print_1_to_N_optimal(n - 1);
        System.out.print(n + " ");
    }

    // Brute Force (N to 1): Pre-recursion print
    public static void print_N_to_1_brute(int n) {
        if (n == 0) {
            return;
        }
        System.out.print(n + " ");
        print_N_to_1_brute(n - 1);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        if (!scanner.hasNextInt()) return;
        int N = scanner.nextInt();
        
        System.out.print("N to 1 (Brute): ");
        print_N_to_1_brute(N);
        System.out.print("\n1 to N (Optimal): ");
        print_1_to_N_optimal(N);
        System.out.println();
        scanner.close();
    }
}

File: day015_pure_recursion/related_print_n_to_1/solution.cpp
#include <iostream>

using namespace std;

// Optimized (1 to N): Post-recursion print
void print_1_to_N_optimal(int n) {
    if (n == 0) {
        return;
    }
    print_1_to_N_optimal(n - 1); // Recurse first (build stack)
    cout << n << " ";           // Print after return (unwind stack)
}

// Brute Force (N to 1): Pre-recursion print
void print_N_to_1_brute(int n) {
    if (n == 0) {
        return;
    }
    cout << n << " ";           // Print first (manage state before call)
    print_N_to_1_brute(n - 1);
}

void solve_print_n() {
    int N;
    if (!(cin >> N)) return;

    cout << "N to 1 (Brute): ";
    print_N_to_1_brute(N);
    cout << "\n1 to N (Optimal): ";
    print_1_to_N_optimal(N);
    cout << endl;
}

int main() {
    solve_print_n();
    return 0;
}